src/util/locking/GreedyReadWriteLocker.ts

```diff
@@ -0,0 +1,122 @@
+import { BasicRepresentation } from '../../ldp/representation/BasicRepresentation';
+import type { ResourceIdentifier } from '../../ldp/representation/ResourceIdentifier';
+import type { ResourceStore } from '../../storage/ResourceStore';
+import { ForbiddenHttpError } from '../errors/ForbiddenHttpError';
+import { NotFoundHttpError } from '../errors/NotFoundHttpError';
+import { readableToString } from '../StreamUtil';
+import type { ReadWriteLocker } from './ReadWriteLocker';
+import type { ResourceLocker } from './ResourceLocker';
+
+export interface GreedyReadWriteSuffixes {
+  count: string;
+  read: string;
+  write: string;
+}
+
+/**
+ * A {@link ReadWriteLocker} that allows for multiple simultaneous read operations.
+ * Write operations will be blocked as long as read operations are not finished.
+ * New read operations are allowed while this is going on, which will cause write operations to wait longer.
+ */
+export class GreedyReadWriteLocker implements ReadWriteLocker {
+  private readonly locker: ResourceLocker;
+  private readonly store: ResourceStore;
```

Comment:
So this should definitely be a more simple interface, like a key/value store or so.
One possible implementation happens to be a `ResourceStore`.